### DISCUSSION TITLE
Modbus Flow control

### DISCUSSION THREAD

**Discussion by werneru** [syssi/esphome-soyosource-gtn-virtual-meter#181] [General]
First of all; great work!!  I use the RS485/TTL converter wIth the DI DE RE RO pins. In the yaml is written that the GPIO012 is used for flow control but it seems to be not working as expected. I would have expect, that this GPIO012 would go to high with send to the Soyo and then directly to low for receiving. DE-RE is bridged. Sending is working but nothing is received. Then I took out the bridge between DE RE and put RE to low. I can now send and receive but with many "header errors". Which pin is actually used for this flow control on D1_Mini? Is it implemented as I expected?

**Comment by syssi** [comment]
Which hardware revision of the Soyosource do you own? Are you sure your Soyosource isn't a silent one? Are you also sure the received data isn't just an echo of your transmission/request? I'm pretty sure the flow control is implemented properly.

**Reply by werneru** [reply]
Thanks for the quick answer...  Yes, i'm sure that this Soyo is sending information out.
<img width="725" alt="Screenshot 2024-07-21 at 10 32 06" src="https://github.com/user-attachments/assets/6a1835e2-4bd2-4a73-a9ce-a51ab3559e00">
<img width="694" alt="Screenshot 2024-07-21 at 10 32 15" src="https://github.com/user-attachments/assets/0a3e4758-f16c-4768-91ac-359f12c4a363">

As i wrote, if I bride DE-RE and connect it to GPIO012 I get no information back from the Soyo but the limiter works. If i remove the bridge and set RE to low then I get information with errors.

**Reply by werneru** [reply]
Imo, the Flow control pin should set to high only if something send to the Soyo. Else it should be low for receiving. Is the GPIO012 correct on D1 Mini?

**Reply by syssi** [reply]
You can use/configure any GPIO for flow control but it needs to be enabled here:

```
soyosource_modbus:
  - id: modbus0
    uart_id: uart_0
    flow_control_pin: GPIO12
    name: firstfloor_inverter
```

**Reply by syssi** [reply]
`GPIO12` of your ESP8266 Wemos D1 mini board is labeled with `D6`. Do you measure a signal here on every `TX`?

**Reply by werneru** [reply]
just compiled a new D1_mini by using D3 for flow control and DE-RE bridged. I works and no header error anymore.
Many thanks!

**Reply by garylankford8** [reply]
I just bought two of these inverters on AlliExpress after finding this project... If I had read more deeply, I would have known about the later production "silent" inverters and not bought them, lol!

Does this thread suggest a path to addressing 'silent' inverters, not explored before, or is the inverter werneru has an older type?

I put together an esp32 with 2 RS485-ttl modules using the 'esp32 multiple inverter example' code and get the error (same as if the modbus plug isn't plugged in), obviously no communication.

```[09:24:00][W][soyosource_inverter:113]: 'gtn_vm_l1': The inverter didn't respond to the last 15 status requests```

**Reply by syssi** [reply]
@garylankford8 Could you create a new discussion or issue? I will provide some guidance / support.
